### DISCUSSION TITLE
Recommended guidelines for application and desktop shortcuts?

### DISCUSSION THREAD

**Discussion by hab25** [manna-harbour/miryoku#225] [Q&A] [answered]
I've recently switched to miryoku (it is a masterpiece; thank you for all your work, @manna-harbour), and since so much about the way I use the keyboard has changed anyway (especially because I'm coming from QWERTY), I'm taking the opportunity to move around my custom keybindings.

Do you (readers) have any recommendations? By the way, @manna-harbour, I would expect this situation to be quite common, and many users may find it helpful to see your answer (I expect it would be of very high quality) featured somewhere in Miryoku's documentation.

An overview of my current usage (feel free to not read, criticism very welcome): 
- completely default Miryoku on a Corne with ZMK
- `control-leftalt-shift-super-<SOME_CHAR>` for all my custom shortcuts; the high amount (four) of mods is to avoid colliding with application defaults as I generally have a preference (which can be reconsidered) for using those as-is. 
- for this mod combination, I still have plenty of unallocated keys and I'm leaving all six thumb keys open so that I may use them as emacs-style prefix keys if I ever need more.
- I considered customizing miryoku so that the keys right under the two index fingers' homes map to the 4 mods when held, but decided against it to avoid colliding with Miryoku in the event that it adds a feature I'm interested in to the holds of those positions. Also, against my initial expectations, holding the 4 mods feels just as comfortable as holding shift now.

**Comment by third774** [comment]
On my MacBook, I have caps lock re-bound (using Karabiner Elements) to be escape on tap and meh (⌃⌥⇧) on hold, then I can also easily hold ⌘ in addition for another layer of shortcuts. These combinations are easy enough with home row mods and miryoku that I have no problem using them, and it keeps all of my shortcuts accessible when using the laptop's built-in keyboard.

**Comment by hab25** [comment]
Hi @manna-harbour, have you considered answering the question
> Do you (readers) have any recommendations?   

If so, but you decided not to answer because there's something bad about the post, please let me know what it is so I may avoid repeating this mistake in the Miryoku forums.

**Reply by manna-harbour** [reply]
Please feel welcome to ask any questions you like (after searching the docs / issues / discussions of course)!

**Comment by manna-harbour** [comment, marked as answer]
Miryoku was designed so that any combination of any mods and any other key can be made with minimum effort, so I just use existing keybindings when available. I don't have many custom keybindings, so I can just use `Super` for those.

Using all four mods seems fine if your switches are light enough. If not, you could replace `AltGr` with `Hyper`. `AltGr` is not on the diagrams, but it's on the key below `Alt` on both sides. The equivalent index finger key could be used instead, except that it's used for [layer lock on sub-layers with a double tap](https://github.com/manna-harbour/miryoku/tree/master/docs/reference#additional-features). It's possible, although unlikely, that it could change to a hold in future, so the ring finger key is safer if you don't need `AltGr`. And also when [mod customisation](https://github.com/manna-harbour/miryoku/issues/33) is supported it'll be easy to replace `AltGr` with another mod such as `Hyper` [via `custom_config.h`](https://github.com/manna-harbour/miryoku/discussions/85).

**Reply by hab25** [reply]
> I just use existing keybindings when available

Question: do you attempt to keep key bindings consistent across applications? E.g., assuming your web browser's buffers are not managed by Emacs: is your keybinding for switching Emacs buffers the same as your keybinding for switching your web browser's buffers?

> I don't have many custom keybindings, so I can just use `Super` for those.

I'm especially glad about this part of your answer as it made me think and decide I'll be switching to `Super` as well. (I realized that, though desktop environments tend to like using `Super`, the risk of colliding with defaults is probably low because they don't provide many: it's hard to do so when workflows can vary so much.)

Also, as one can guess from the previous paragraph, I've since changed my mind and this statement is not true: 
>> holding the 4 mods feels just as comfortable as holding shift now

> Using all four mods seems fine if your switches are light enough.

Yes, they are a bit too heavy right now (but not intolerably so); I would guess that light switches work better for Miryoku in general since more holding is involved. I think the next switches I'll be experimenting with will be the lightest high-quality low-profile switches I can find. 

> If not, you could replace AltGr with Hyper

I speak a language where the AltGr is helpful, so I discarded this.

**Reply by manna-harbour** [reply]
No, I don't really try, but I do consider it periodically! I'd probably do it via an application aware translator, to avoid having to change the keybindings in the applications.

**Comment by hab25** [comment]
I was previously simply using Super, but I have decided to place all my custom keybindings instead within this constraint from Emacs's conventions for user-reserved key sequences: https://github.com/emacs-mirror/emacs/blob/c97e7a2da2e5707fc94c2c5e2ddd5f2395cdb80b/doc/lispref/tips.texi#L286C1-L288C56.

I believe this should further decrease the risk of collision with defaults, because:
- Desktop environments often by default use Super.
- In new GUI applications for Linux, AFAIU it is practically standard that by default both `ctrl-c` and `ctrl-insert` issue a "copy to clipboard". See also https://en.wikipedia.org/wiki/IBM_Common_User_Access#:~:text=The%20subset%20of%20CUA%20implemented%20in%20Microsoft%20Windows%20or%20OSF/Motif%20is%20generally%20considered%20a%20de%20facto%20standard%20to%20be%20followed%20by%20any%20new%20Unix%20GUI%20environment.
- For such copying, I exclusively use "Copy" on Miryoku's Nav layer, which produces "ctrl-insert" in my (the default) case; so I'd never use `ctrl-c` otherwise.
- I use https://github.com/xremap/xremap which supports arbitrary-length key sequences and graphical-application-specific mapping, so AFAICT this scheme can be supported with any graphical application.
